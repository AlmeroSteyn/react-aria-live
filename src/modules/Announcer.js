import PropTypes from 'prop-types';
import React, { Component } from 'react';
import MessageBlock from './MessageBlock';

class Announcer extends Component {
  state = {
    assertiveMessage1: '',
    assertiveMessage2: '',
    politeMessage1: '',
    politeMessage2: '',
    oldPolitemessage: '',
    oldAssertiveMessage: '',
    setAlternatePolite: false,
    setAlternateAssertive: false,
  };

  static propTypes = {
    politeMessage: PropTypes.string,
    assertiveMessage: PropTypes.string,
  };

  static getDerivedStateFromProps(nextProps, state) {
    const { oldPolitemessage, oldAssertiveMessage } = state;
    const { politeMessage, assertiveMessage } = nextProps;

    if (oldPolitemessage !== politeMessage) {
      return {
        politeMessage1: state.setAlternatePolite ? '' : politeMessage,
        politeMessage2: state.setAlternatePolite ? politeMessage : '',
        oldPolitemessage: politeMessage,
        setAlternatePolite: !state.setAlternatePolite,
      };
    }

    if (oldAssertiveMessage !== assertiveMessage) {
      return {
        assertiveMessage1: state.setAlternateAssertive ? '' : assertiveMessage,
        assertiveMessage2: state.setAlternateAssertive ? assertiveMessage : '',
        oldAssertiveMessage: assertiveMessage,
        setAlternateAssertive: !state.setAlternateAssertive,
      };
    }

    return null;
  }

  render() {
    const {
      assertiveMessage1,
      assertiveMessage2,
      politeMessage1,
      politeMessage2,
    } = this.state;
    return (
      <div>
        <MessageBlock aria-live="assertive" message={assertiveMessage1} />
        <MessageBlock aria-live="assertive" message={assertiveMessage2} />
        <MessageBlock aria-live="polite" message={politeMessage1} />
        <MessageBlock aria-live="polite" message={politeMessage2} />
      </div>
    );
  }
}

export default Announcer;
