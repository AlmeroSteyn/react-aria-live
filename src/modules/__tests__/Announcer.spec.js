import React from 'react';
import { mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';
import Announcer from '../Announcer';

describe('Announcer', () => {
  it('should render correctly via snapshot', () => {
    const wrapper = mount(<Announcer assertiveMessage="" politeMessage="" />);

    const tree = mountToJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it('should alternate assertive messages', () => {
    const wrapper = mount(<Announcer assertiveMessage="" politeMessage="" />);

    wrapper.setProps({ assertiveMessage: 'I am a message' });

    expect(wrapper.state()).toEqual({
      assertiveMessage1: 'I am a message',
      assertiveMessage2: '',
      oldAssertiveMessage: 'I am a message',
      oldPolitemessage: '',
      politeMessage1: '',
      politeMessage2: '',
      setAlternateAssertive: true,
      setAlternatePolite: false,
    });
    expect(wrapper.find('MessageBlock').get(0).props.message).toBe(
      'I am a message'
    );
    expect(wrapper.find('MessageBlock').get(1).props.message).toBe('');

    wrapper.setProps({ assertiveMessage: 'I am a changed message' });

    expect(wrapper.state()).toEqual({
      assertiveMessage1: '',
      assertiveMessage2: 'I am a changed message',
      oldAssertiveMessage: 'I am a changed message',
      oldPolitemessage: '',
      politeMessage1: '',
      politeMessage2: '',
      setAlternateAssertive: false,
      setAlternatePolite: false,
    });
    expect(wrapper.find('MessageBlock').get(0).props.message).toEqual('');
    expect(wrapper.find('MessageBlock').get(1).props.message).toEqual(
      'I am a changed message'
    );
  });

  it('should alternate polite messages', () => {
    const wrapper = mount(<Announcer assertiveMessage="" politeMessage="" />);

    wrapper.setProps({ politeMessage: 'I am a message' });

    expect(wrapper.state()).toEqual({
      assertiveMessage1: '',
      assertiveMessage2: '',
      oldAssertiveMessage: '',
      oldPolitemessage: 'I am a message',
      politeMessage1: 'I am a message',
      politeMessage2: '',
      setAlternateAssertive: false,
      setAlternatePolite: true,
    });
    expect(wrapper.find('MessageBlock').get(2).props.message).toBe(
      'I am a message'
    );
    expect(wrapper.find('MessageBlock').get(3).props.message).toBe('');

    wrapper.setProps({ politeMessage: 'I am a changed message' });

    expect(wrapper.state()).toEqual({
      assertiveMessage1: '',
      assertiveMessage2: '',
      oldAssertiveMessage: '',
      oldPolitemessage: 'I am a changed message',
      politeMessage1: '',
      politeMessage2: 'I am a changed message',
      setAlternateAssertive: false,
      setAlternatePolite: false,
    });
    expect(wrapper.find('MessageBlock').get(2).props.message).toEqual('');
    expect(wrapper.find('MessageBlock').get(3).props.message).toEqual(
      'I am a changed message'
    );
  });
});
