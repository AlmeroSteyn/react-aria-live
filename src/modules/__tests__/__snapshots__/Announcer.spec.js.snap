// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Announcer should render correctly via snapshot 1`] = `
<Announcer
  assertiveMessage=""
  politeMessage=""
>
  <div>
    <MessageBlock
      aria-live="assertive"
      message=""
    >
      <div
        aria-live="assertive"
        role="log"
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "width": "1px",
          }
        }
      />
    </MessageBlock>
    <MessageBlock
      aria-live="assertive"
      message=""
    >
      <div
        aria-live="assertive"
        role="log"
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "width": "1px",
          }
        }
      />
    </MessageBlock>
    <MessageBlock
      aria-live="polite"
      message=""
    >
      <div
        aria-live="polite"
        role="log"
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "width": "1px",
          }
        }
      />
    </MessageBlock>
    <MessageBlock
      aria-live="polite"
      message=""
    >
      <div
        aria-live="polite"
        role="log"
        style={
          Object {
            "border": 0,
            "clip": "rect(0 0 0 0)",
            "height": "1px",
            "margin": "-1px",
            "overflow": "hidden",
            "padding": 0,
            "position": "absolute",
            "width": "1px",
          }
        }
      />
    </MessageBlock>
  </div>
</Announcer>
`;
